msgraph/core/middleware/retry.py

```diff
@@ -0,0 +1,223 @@
+        retry_total takes precedence over all other counts.
+        """
+        return cls(max_retries=0)
+
+    def get_retry_options(self, middleware_control):
+        """
+        Check if request specific configs have been passed and override any session defaults
+        Then configure retry settings into the form of a dict.
+        """
+        if middleware_control:
+            return {
+                'total':
+                min(middleware_control.get('max_retries', self.max_retries), self.MAX_RETRIES),
+                'backoff':
+                middleware_control.get('retry_backoff_factor', self.backoff_factor),
+                'max_backoff':
+                middleware_control.get('retry_backoff_max', self.backoff_max),
+                'timeout':
+                middleware_control.get('retry_time_limit', self.timeout),
+                'retry_codes':
+                set(middleware_control.get('retry_on_status_codes', self._retry_on_status_codes))
+                | set(self._DEFAULT_RETRY_STATUS_CODES),
+                'methods':
+                self._allowed_methods,
+            }
+        return {
+            'total': self.max_retries,
+            'backoff': self.backoff_factor,
+            'max_backoff': self.backoff_max,
+            'timeout': self.timeout,
+            'retry_codes': self._retry_on_status_codes,
+            'methods': self._allowed_methods,
+        }
+
+    def send(self, request, **kwargs):
+        """
+        Sends the http request object to the next middleware or retries the request if necessary.
+        """
+        retry_options = self.get_retry_options(request.context.middleware_control)
+        absolute_time_limit = min(retry_options['timeout'], self.MAX_DELAY)
+        response = None
+        retry_count = 0
+        retry_valid = True
+
+        while retry_valid:
+            try:
+                start_time = time.time()
+                request.headers.update({'retry-attempt': '{}'.format(retry_count)})
+                response = super().send(request, **kwargs)
+                # Check if the request needs to be retried based on the response method
+                # and status code
+                if self.should_retry(retry_options, response):
+                    # check that max retries has not been hit
+                    retry_valid = self.check_retry_valid(retry_options, retry_count)
+
+                    # Get the delay time between retries
+                    delay = self.get_delay_time(retry_options, retry_count, response)
+
+                    if retry_valid and delay < absolute_time_limit:
+                        time.sleep(delay)
+                        end_time = time.time()
+                        absolute_time_limit -= (end_time - start_time)
+                        # increment the count for retries
+                        retry_count += 1
+
+                        continue
+                break
+            except Exception as error:
```

Comment:
We aren't expecting a specific exception here so we should let this exception be handled higher up.

Reply:
Updated accordingly.